README.md
# rollup-plugin-analyzer

> Mad metrics for your rollup bundles, know all the things

[![npm](https://img.shields.io/npm/v/rollup-plugin-analyzer?style=for-the-badge)](https://npmjs.org/package/rollup-plugin-analyzer)   [![Travis (.com)](https://img.shields.io/travis/com/doesdev/rollup-plugin-analyzer?style=for-the-badge)](https://travis-ci.com/doesdev/rollup-plugin-analyzer)   [![Dependents (via libraries.io)](https://img.shields.io/librariesio/dependents/npm/rollup-plugin-analyzer?style=for-the-badge)](https://github.com/doesdev/rollup-plugin-analyzer/network/dependents)   [![npm](https://img.shields.io/npm/dm/rollup-plugin-analyzer?style=for-the-badge)](https://npmcharts.com/compare/rollup-plugin-analyzer?minimal=true&interval=7)

## TOC

- [Install](#install)
- [Usage](#usage)
  + [Importing or Requiring](#importing-or-requiring)
    - [Import as ES Module](#import-as-es-module)
    - [Requiring as CJS](#requiring-as-cjs)
  + [Usage from rollup config](#usage-from-rollup-config)
  + [Usage from build script](#usage-from-build-script)
  + [CI usage example](#ci-usage-example)
  + [Example results](#results)
  + [Example results (with `summaryOnly` enabled)](#results--with--summaryonly--enabled-)
- [Options](#options)
- [FAQ](#faq)
- [License](#license)

## Install

```sh
$ npm install --save-dev rollup-plugin-analyzer
```

## Usage

### Importing or Requiring

#### Import as ES Module
```js
import analyze from 'rollup-plugin-analyzer'
```

#### Requiring as CJS
```js
const analyze = require('rollup-plugin-analyzer')
```

### Usage from rollup config
```js
export default {
  entry: 'module.js',
  dest: 'index.js',
  format: 'cjs',
  plugins: [analyze()]
}
```

### Usage from build script
```js
rollup({
  entry: 'main.js',
  plugins: [analyze()]
}).then(...)
```

### CI usage example
```js
const limitBytes = 1e6

const onAnalysis = ({ bundleSize }) => {
  if (bundleSize < limitBytes) return
  console.log(`Bundle size exceeds ${limitBytes} bytes: ${bundleSize} bytes`)
  return process.exit(1)
}

rollup({
  entry: 'main.js',
  plugins: [analyze({ onAnalysis, skipFormatted: true })]
}).then(...)
```

### results
logged to console on rollup completion
```sh
-----------------------------
Rollup File Analysis
-----------------------------
bundle size:    2.809 KB
original size:  11.436 KB
code reduction: 75.44 %
module count:   5

█████████████████████████████████████████████░░░░░
file:            /virtual-insanity.js
bundle space:    90.64 %
rendered size:   2.546 KB
original size:   2.57 KB
code reduction:  0.93 %
dependents:      1
  - /jamiroquai.js

██░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
file:            /bundle-a.js
bundle space:    4.27 %
rendered size:   120 Bytes
original size:   309 Bytes
code reduction:  61.17 %
dependents:      0

█░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
file:            /jamiroquai.js
bundle space:    2.95 %
rendered size:   83 Bytes
original size:   169 Bytes
code reduction:  50.89 %
dependents:      1
  - /the-alphabet-but-incomplete.js
...
```

### results (with `summaryOnly` enabled)
```sh
-----------------------------
Rollup File Analysis
-----------------------------
bundle size:    2.809 KB
original size:  11.436 KB
code reduction: 75.44 %
module count:   5

/virtual-insanity.js
█████████████████████████████████████████████░░░░░ 90.64 % (2.546 KB)
/bundle-a.js
██░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 4.27 % (120 Bytes)
/jamiroquai.js
█░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 2.95 % (83 Bytes)
/the-alphabet-but-incomplete.js
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 1.17 % (33 Bytes)
/the-declaration-of-independence.js
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 0.96 % (27 Bytes)
```

## Options

- **stdout** - *optional*
  - type: Boolean
  - default: `false`
  - description: Print to stdout (console.log) instead of stderr (console.error)
- **limit** - *optional*
  - type: Number
  - default: `null`
  - description: Limit number of files to output analysis of, sorted by DESC size
- **filter** - *optional*
  - type: Array | String | Function
  - default: `null`
  - description: Filter module to output in analysis results.
    - If a string is passed, checks if module name contains the string
    - If array it checks the same for each string in the array
    - If function, return a boolean to indicate if a module should be in analysis results
    - This only suppresses the modules details, it does not affect the summary output
    - If you would like it to filter from summary info as well set `filterSummary` to `true`
  - notes: Function receives `module` object specified below, should return boolean
- **filterSummary** - *optional*
  - type: Boolean
  - default: `false`
  - description: If `true` the `filter` and `limit` options will also remove any filtered out module data from the summary
- **root** - *optional*
  - type: String
  - default: `process.cwd()`
  - description: Application directory, used to display file paths relatively
- **hideDeps** - *optional*
  - type: Boolean
  - default: `false`
  - description: Don't itemize dependents in the formatted output
- **showExports** - *optional*
  - type: Boolean
  - default: `false`
  - description: Show used and unused exports
- **summaryOnly** - *optional*
  - type: Boolean
  - default: `false`
  - description: Only output bundle summary and module usage bar graphs
- **skipFormatted** - *optional*
  - type: Boolean
  - default: `false`
  - description: Don't output formatted string
- **writeTo** - *optional*
  - type: Function
  - default: `null`
  - description: Callback to be invoked with formatted string
  - function will be invoked with:
    - **analysisString** *(String)*
- **transformModuleId** - *optional*
  - type: Function
  - default: `null`
  - description: Modify module ids
  - function will be invoked with:
    - **id** *(String)* - path of module / rollup module id
  - function should return:
    - **id** *(String)* - desired final module id to display in analysis results
  - example: `(id) => id.replace(/^\0(?:commonjs-proxy:)?/, '')`
- **onAnalysis** - *optional*
  - type: Function
  - default: `null`
  - description: Callback to be invoked with analysis object
  - function will be invoked with:
    - **analysisObject** *(Object)*
      - **bundleSize** *(Number)* - rendered bundle size in bytes
      - **bundleOrigSize** *(Number)* - original bundle size in bytes
      - **bundleReduction** *(Number)* - percentage of rendered bundle size reduction
      - **moduleCount** *(Number)* - Count of all included modules
      - **modules** *(Array)* - array of `module` analysis objects
        - **module** *(Object)*
          - **id** *(String)* - path of module / rollup module id
          - **size** *(Number)* - size of rendered module code in bytes
          - **origSize** *(Number)* - size of module's original code in bytes
          - **dependents** *(Array)* - list of dependent module ids / paths
          - **percent** *(Number)* - percentage of module size relative to entire bundle
          - **reduction** *(Number)* - percentage of rendered size reduction
          - **renderedExports** *(Array)* - list of used named exports
          - **removedExports** *(Array)* - list of unused named exports

## FAQ

#### Why is the reported size not the same as the file on disk?

This module is geared towards the details of the individual modules that make up the bundle and their relative impact to bundle size. That's a detailed way of saying, it doesn't really care about size on disk. There are other options which focus on size on disk as well as delivery size which can be used alongside this module (or in place of if your concern is not per module impact). In particular [rollup-plugin-size-snapshot](https://github.com/TrySound/rollup-plugin-size-snapshot) seems like a great option for that.

Getting a bit further into the details, rather than just intent, of why the reported size differs from that on disk. We get the module data from Rollup which reports it after chunk (module) resolution and tree-shaking, but before post-processing (such as minification and compression). We then add the sizes of each of those modules together, this is the `bundle size` that we report.

That means it won't account for post-processing from other plugins and also won't account for post-processing by Rollup itself, which includes boilerplate / shims depending on what the output format is (CJS, ESM, iife, etc...).

#### Why am I seeing multiple analysis outputs emitted?

Rollup allows you to output to multiple files. If you are outputting to multiple files you will get a distinct analysis for each output file. Each analysis will contain data on the files imported by the respective target.

## License

MIT © [Andrew Carpenter](https://github.com/doesdev)


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

### UNRELEASED

> 10 July 2021

- Update dev deps
- Remove husky
- Specify exports = default in rollup config
- Add 2.0.x tests, remove 1.0.x tests

### [4.0.0](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.3.0...4.0.0)

> 19 December 2020

- **[BREAKING]** Limit does not affect summary (unless `filterSummary === true`)
- [DEV] Update dev dependencies (standard, husky)
- [DEV] Fix standard style issue
- [DOCS] Add FAQ to readme
- [DOCS] Add TOC to readme

### [3.3.0](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.2.3...3.3.0)

> 8 July 2020

- Add `filterSummary` option (#19 && #20)
- Fix `origSize`, `bundleOrigSize`, and `bundleReduction` (was apparently borked)
- Remove Rollup 0.60.0 from tests
- Fix incorrect key names in `analysisObject` docs
- Clarify `filter` usage
- Update dev deps

### [3.2.3](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.2.2...3.2.3)

> 13 May 2020

- [Add TS declarations](https://github.com/doesdev/rollup-plugin-analyzer/pull/17)
- Update lockfile

### [3.2.2](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.2.1...3.2.2)

> 1 November 2019

- Hack: Coalesce bundleMods
- Update dev deps

### [3.2.1](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.2.0...3.2.1)

> 12 September 2019

- Explicitly add `.default` property to `plugin` in CJS output
- Because my previous unbreaking change broke things
- Clarify import and require usage in README

### [3.2.0](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.1.2...3.2.0)

> 12 September 2019

- [SHOULD NOT BE BREAKING] Correct export signature in module.js
  - This, while an API change, should not be a breaking change
  - Previously the export signature when used as a module was a bit effed
  - This should fix that
  - To my knowledge and in my tests that should not break things
- Update yarn lockfile using latest Rollup version
- Correct path to latest Rollup version under scripts
- Remove `changelog` script (manually maintain changelog, tis better this way)
- Add `engines` to package.json, Node >= 8
- Update dev dependencies (mvt, husky, standard)
- Fix some standard style issues

### [3.1.2](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.1.1...3.1.2)

> 14 July 2019

- Update dependencies
- Fix some standardjs issues
- Use mvt for testing instead of AVA

### [3.1.1](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.1.0...3.1.1)

> 7 July 2019

- Don't combine bar graph with id in summary

### [3.1.0](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.0.1...3.1.0)

> 6 July 2019

- Add bar graphs to formatted output
- Add `summaryOnly` option

### [3.0.1](https://github.com/doesdev/rollup-plugin-analyzer/compare/3.0.0...3.0.1)

> 1 June 2019

- Update dependencies

### [3.0.0](https://github.com/doesdev/rollup-plugin-analyzer/compare/2.1.0...3.0.0)

> 5 January 2019

- Add default export (fixes #6), warn on rollup < 1.0.0 [`#6`](https://github.com/doesdev/rollup-plugin-analyzer/issues/6)
- Remove support for rollup < 1.0.0, support 1.0.0 [`506d93f`](https://github.com/doesdev/rollup-plugin-analyzer/commit/506d93f942dce5e076312bd1642aed182fe12a60)
- Implement transformModuleId option [`a173a43`](https://github.com/doesdev/rollup-plugin-analyzer/commit/a173a43f9e6b7a5fee8c07274e7d43df7a6939db)
- Update Node versions to test in CI, drop 6 [`ee60a38`](https://github.com/doesdev/rollup-plugin-analyzer/commit/ee60a3868b946760560c71a0d594a32965073762)

#### [2.1.0](https://github.com/doesdev/rollup-plugin-analyzer/compare/2.0.4...2.1.0)

> 11 July 2018

- Let onAnalysis and formatted coexist, addresses #2 [`0bdc851`](https://github.com/doesdev/rollup-plugin-analyzer/commit/0bdc851dd66623584e4ac0f41636d0f30340a6e6)
- Don't swallows errors, as noted in #1 [`07f5c9a`](https://github.com/doesdev/rollup-plugin-analyzer/commit/07f5c9a15e58b49004df8aa65d757704d5959609)
- Add test to ensure plugin isn't adding overhead [`513bd8b`](https://github.com/doesdev/rollup-plugin-analyzer/commit/513bd8bb44388b9d09c15f79f0b09b1ffdf1cfb6)
- Add CONTRIBUTING.md [`77ed765`](https://github.com/doesdev/rollup-plugin-analyzer/commit/77ed7650e1140cf8bba89722b304ca00a0c7c804)
- Test that callback is only invoked once, failing [`8c6ac14`](https://github.com/doesdev/rollup-plugin-analyzer/commit/8c6ac148fdf74b5beadd7b6f8f17a847fb2572bd)
- Bump to 2.1.0 [`3fa544f`](https://github.com/doesdev/rollup-plugin-analyzer/commit/3fa544fd823b1fb38797b151a1485ab5c7c8e8c4)

#### [2.0.4](https://github.com/doesdev/rollup-plugin-analyzer/compare/2.0.3...2.0.4)

> 3 July 2018

- Update multi-out tests, failing, skipped [`8813ec7`](https://github.com/doesdev/rollup-plugin-analyzer/commit/8813ec7786825b1256c9af9331f55941e1ad718e)
- Handle bundles with multiple outputs, 2.0.4 [`12512c9`](https://github.com/doesdev/rollup-plugin-analyzer/commit/12512c9b2a7d2f2c61467b295add5286a79abae2)
- Test with experimentalCodeSplitting enabled as well [`fc54b1b`](https://github.com/doesdev/rollup-plugin-analyzer/commit/fc54b1be5ed8b360ce30cdf511f5671a8f974923)
- Fix module name in examples [`dafa44e`](https://github.com/doesdev/rollup-plugin-analyzer/commit/dafa44edb863d6e5d13953a11aa79e920d989158)

#### [2.0.3](https://github.com/doesdev/rollup-plugin-analyzer/compare/2.0.2...2.0.3)

> 9 June 2018

- Add showExports option, expose exports in analysis [`2be9c1c`](https://github.com/doesdev/rollup-plugin-analyzer/commit/2be9c1c9f82d7a01d3e638971b2127c363b9cfc9)
- Some cleanup and formatting tweaks [`e40789e`](https://github.com/doesdev/rollup-plugin-analyzer/commit/e40789eb1047d0ed04d30b5254a9373034453325)
- Add module count to formatted output [`e880a9d`](https://github.com/doesdev/rollup-plugin-analyzer/commit/e880a9d2fca46777199017abb83c133878c20c5b)

html/report.js
'use strict'

const d3 = window.d3
const testEl = document.getElementById('size-test-el')
const [winX, winY] = [testEl.offsetWidth, testEl.offsetHeight]
const maxSize = winX < winY ? winX : winY
const width = maxSize - 20
const height = maxSize - 20
const radius = Math.min(width, height) / 2
const color = d3.scaleOrdinal(d3.schemeCategory20b)

const analysis = {
  bundleSize: 2809,
  bundleOrigSize: 11436,
  bundleReduction: 75.44,
  modules: [
    {
      id: '\\virtual-insanity.js',
      size: 2546,
      origSize: 2570,
      renderedExports: ['virtualInsanity'],
      removedExports: [],
      dependents: ['\\jamiroquai.js'],
      percent: 90.64,
      reduction: 0.93
    },
    {
      id: '\\bundle-a.js',
      size: 120,
      origSize: 309,
      renderedExports: [],
      removedExports: [],
      dependents: [],
      percent: 4.27,
      reduction: 61.17
    },
    {
      id: '\\jamiroquai.js',
      size: 83,
      origSize: 169,
      renderedExports: ['smallNestedConstA', 'largeNestedConstB'],
      removedExports: [],
      dependents: ['\\the-alphabet-but-incomplete.js'],
      percent: 2.95,
      reduction: 50.89
    },
    {
      id: '\\the-alphabet-but-incomplete.js',
      size: 33,
      origSize: 123,
      renderedExports: ['anotherSmallConst'],
      removedExports: [],
      dependents: ['\\bundle-a.js'],
      percent: 1.17,
      reduction: 73.17
    },
    {
      id: '\\the-declaration-of-independence.js',
      size: 27,
      origSize: 8265,
      renderedExports: ['aSmallConst'],
      removedExports: ['aLargeConst'],
      dependents: ['\\bundle-a.js'],
      percent: 0.96,
      reduction: 99.67
    }
  ],
  moduleCount: 5
}

let rootModule
const depTree = {}
analysis.modules.forEach((m) => {
  if (!m.dependents.length) rootModule = m
  m.dependents.forEach((d) => (depTree[d] = depTree[d] || []).push(m))
})

const data = {
  name: rootModule.id,
  children: []
}

const addChild = (parent, m) => {
  const { id: name, size } = m
  const tree = depTree[name]
  const el = tree ? { name, children: [] } : { name, size }
  parent.children = parent.children || []
  parent.children.push(el)

  if (tree) tree.forEach((d) => addChild(el, d))

  return el
}

depTree[data.name].forEach((d) => addChild(data, d))

const renderSunburst = (nodeData) => {
  const svg = d3.select('svg')
  svg.attr('width', width)
  svg.attr('height', height)

  const g = svg.append('g')
  g.attr('transform', `translate(${width / 2}, ${height / 2})`)

  const partition = d3.partition().size([2 * Math.PI, radius])
  const root = d3.hierarchy(nodeData).sum((d) => d.size)

  partition(root)

  const arc = d3.arc()
  arc.startAngle((d) => d.x0)
  arc.endAngle((d) => d.x1)
  arc.innerRadius((d) => d.y0)
  arc.outerRadius((d) => d.y1)

  const applyRootViz = () => {
    (g.selectAll('g')
      .data(root.descendants())
      .enter()
      .append('g').attr('class', 'node')
      .append('path')
      .attr('display', (d) => d.depth ? null : 'none')
      .attr('d', arc)
      .style('stroke', '#fff')
      .style('fill', (d) => {
        return color((d.children ? d : d.parent).data.name)
      })
    )
  }

  applyRootViz()

  const computeTextRotation = (d) => {
    const angle = (d.x0 + d.x1) / Math.PI * 90
    return (angle < 90 || angle > 270) ? angle : angle + 180
  }

  const applyNodesViz = () => {
    (g.selectAll('.node')
      .append('text')
      .attr('transform', (d) => {
        const translate = arc.centroid(d)
        const rotate = computeTextRotation(d)
        return `translate(${translate}), rotate(${rotate})`
      })
      .attr('dx', '-20')
      .attr('dy', '.5em')
      .text((d) => d.parent ? d.data.name : '')
    )
  }

  applyNodesViz()
}

renderSunburst(data)
